core/services/ocr2/plugins/ocr2keeper/evm/mercury_lookup.go

```diff
@@ -0,0 +1,343 @@
+package evm
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/patrickmn/go-cache"
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
+)
+
+type MercuryLookup struct {
+	feedLabel  string
+	feeds      []string
+	queryLabel string
+	query      *big.Int
+	extraData  []byte
+}
+
+// MercuryMultiResponse TODO guessing on this
+type MercuryMultiResponse struct {
+	ChainlinkBlobs []string `json:"chainlinkBlob"`
+}
+
+type MercuryResponse struct {
+	ChainlinkBlob string `json:"chainlinkBlob"`
+}
+
+type MercuryBytes struct {
+	Index int
+	Error error
+	Bytes []byte
+}
+
+const (
+	RetryDelay   = 500 * time.Millisecond
+	TotalAttempt = 3
+)
+
+// mercuryLookup looks through check upkeep results looking for any that need off chain lookup
+func (r *EvmRegistry) mercuryLookup(ctx context.Context, upkeepResults []types.UpkeepResult) ([]types.UpkeepResult, error) {
+	// return error only if there are errors which stops the process
+	// don't surface Mercury API errors to plugin bc MercuryLookup process should be self-contained
+	for i := range upkeepResults {
+		// if its another reason continue/skip
+		if upkeepResults[i].FailureReason != UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED {
+			continue
+		}
+
+		block, upkeepId, err := blockAndIdFromKey(upkeepResults[i].Key)
+		if err != nil {
+			r.lggr.Error("[MercuryLookup] error getting block and upkeep id:", err)
+			return nil, err
+		}
+
+		// checking if this upkeep is in cooldown from api errors
+		_, onIce := r.mercury.cooldownCache.Get(upkeepId.String())
+		if onIce {
+			r.lggr.Infof("[MercuryLookup] cooldown Skipping UpkeepId %s at block %s\n", upkeepId.String(), block.String())
+			continue
+		}
+
+		// if it doesn't decode to the mercury custom error continue/skip
+		mercuryLookup, err := r.decodeMercuryLookup(upkeepResults[i].PerformData)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_MERCURY_LOOKUP_ERROR
```

Comment:
I'm not sure of the purpose of replacing the failure reason here?

will this be replaced for all check reverts, even those which are not ever intended to be mercury upkeeps?

Reply:
updated all to `UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED`